src/custom-apps/components/PermissionAlert/utils.ts

```diff
@@ -0,0 +1,145 @@
+import { gql } from "@apollo/client";
+import { FieldNode, parse, SelectionNode, visit } from "graphql";
+
+interface IntrospectionNode {
+  kind: string;
+  name: string;
+  description: string;
+  fields: Array<{
+    name: string;
+    description: string;
+  }>;
+}
+
+interface PermissionChildNode {
+  permissions: string[];
+  children: string[];
+}
+
+interface PermissionNode {
+  permissions: string[];
+  children: Record<string, PermissionChildNode>;
+}
+
+type PermissionMap = Record<string, PermissionNode>;
+
+// cannot be in `queries.ts` as codegen cannot handle `__schema`
+export const IntrospectionQuery = gql`
+  query PermissionIntrospection {
+    __schema {
+      types {
+        kind
+        name
+        description
+        fields(includeDeprecated: false) {
+          name
+          description
+        }
+      }
+    }
+  }
+`;
+
+const uniq = <T>(value: T, index: number, self: T[]) =>
+  self.indexOf(value) === index;
+
+const extractPermissions = (description?: string) => {
+  const match = (description || "").match(/following permissions(.*): (.*?)\./);
+  const permissions = match ? match[2].split(",") : [];
+
+  return permissions;
+};
```

Comment:
**suggestion:** It's unclear to me how this function extracts permissions, I guess it's based on premise that `description` always follows some regex pattern? Are you sure this pattern will stay the same in future minor & patch releases? I would appreciate a comment here or unit test which provides examples of `description` and corresponding `permissions`.

Reply:
@Droniu a comment along with some unit tests added